llama_index/callbacks/finetuning_handler.py

```diff
@@ -0,0 +1,103 @@
+import json
+from typing import Any, Dict, List, Optional
+
+from llama_index.callbacks.base import BaseCallbackHandler
+from llama_index.callbacks.schema import CBEventType, EventPayload
+
+
+class OpenAIFineTuningHandler(BaseCallbackHandler):
+    def __init__(
+        self,
+    ) -> None:
+        """Initialize the base callback handler."""
+        super().__init__(
+            event_starts_to_ignore=[],
+            event_ends_to_ignore=[],
+        )
+        self._finetuning_events: Dict[str, List[Any]] = {}
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        """Run when an event starts and return id of event."""
+        from llama_index.llms.base import ChatMessage, MessageRole
+
+        if event_type == CBEventType.LLM:
+            cur_messages = []
+            if payload and EventPayload.PROMPT in payload:
+                message = ChatMessage(
+                    role=MessageRole.USER, text=str(payload[EventPayload.PROMPT])
+                )
+                cur_messages = [message]
+            elif payload and EventPayload.MESSAGES in payload:
+                cur_messages = payload[EventPayload.MESSAGES]
+
+            if len(cur_messages) > 0:
+                if event_id in self._finetuning_events:
+                    self._finetuning_events[event_id].extend(cur_messages)
+                else:
+                    self._finetuning_events[event_id] = cur_messages
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        """Run when an event ends."""
+        from llama_index.llms.base import ChatMessage, MessageRole
+
+        if (
+            event_type == CBEventType.LLM
+            and event_id in self._finetuning_events
+            and payload is not None
+        ):
+            if isinstance(payload[EventPayload.RESPONSE], str):
+                response = ChatMessage(
+                    role=MessageRole.ASSISTANT, text=str(payload[EventPayload.RESPONSE])
+                )
+            else:
+                response = payload[EventPayload.RESPONSE].message
+
+            self._finetuning_events[event_id].append(response)
+
+        return None
+
+    def get_finetuning_events(self) -> List[Dict[str, Any]]:
+        events = []
+        for event in self._finetuning_events.values():
+            events.append({"messages": event[:-1], "response": event[-1]})
+
+        return events
+
+    def save_finetuning_events(self, path: str) -> None:
```

Comment:
nit: might be nice to have a docstring since this will end up in API referneces

Reply:
Thanks! Done -- also updated the callbacks docs slightly